File: src/mask.py
#!/usr/bin/env python

import matplotlib.pyplot as plt
import rospy
import ros_numpy
from sensor_msgs.msg import PointCloud2, PointField, Image
import sensor_msgs.point_cloud2 as pcl2
import std_msgs.msg

import struct
import time
import cv2
import numpy as np
import math

import time
import pickle as pkl

import message_filters
from sklearn.neighbors import NearestNeighbors
import open3d as o3d
from scipy import ndimage

def callback (rgb, depth, pc):

    proj_matrix = np.array([[918.359130859375,              0.0, 645.8908081054688, 0.0], \
                            [             0.0, 916.265869140625,   354.02392578125, 0.0], \
                            [             0.0,              0.0,               1.0, 0.0]])

    # process rgb image
    cur_image = ros_numpy.numpify(rgb)
    # cur_image = cv2.cvtColor(cur_image.copy(), cv2.COLOR_BGR2RGB)
    hsv_image = cv2.cvtColor(cur_image.copy(), cv2.COLOR_RGB2HSV)

    # color thresholding
    # --- blue ---
    lower = (98, 100, 100)
    upper = (130, 255, 255)
    mask = cv2.inRange(hsv_image, lower, upper)

    # --- green ---
    lower = (85, 130, 60)
    upper = (95, 255, 255)
    mask_green = cv2.inRange(hsv_image, lower, upper).astype('uint8')

    # bmask = mask.copy() # for checking visibility, max = 255
    # mask = cv2.cvtColor(mask.copy(), cv2.COLOR_GRAY2BGR).astype('uint8')

    # test
    mask = cv2.bitwise_or(mask.copy(), mask_green.copy()) # mask_green.copy()
    mask = cv2.cvtColor(mask.copy(), cv2.COLOR_GRAY2RGB)

    # blob detection
    params = cv2.SimpleBlobDetector_Params()
    # Filter by Color
    params.filterByColor = False
    # Filter by Area.
    params.filterByArea = True
    # Filter by Circularity
    params.filterByCircularity = False
    # Filter by Inerita
    params.filterByInertia = True
    # Filter by Convexity
    params.filterByConvexity = False

    # # Create a detector with the parameters
    # detector = cv2.SimpleBlobDetector_create(params)
    # keypoints = detector.detect(mask)

    # # Find blob centers in the image coordinates
    # blob_image_center = []
    # num_blobs = len(keypoints)
    # for i in range(num_blobs):
    #     blob_image_center.append((keypoints[i].pt[0],keypoints[i].pt[1]))

    # blob_image_center = np.array(blob_image_center)

    # us = blob_image_center[:, 0].astype(int)
    # vs = blob_image_center[:, 1].astype(int)

    # tracking_img = cur_image.copy()
    # for i in range (len(blob_image_center)):
    #     # draw circle
    #     uv = (us[i], vs[i])
    #     cv2.circle(tracking_img, uv, 5, (255, 150, 0), -1)

        # # draw line
        # if i != len(blob_image_center)-1:
        #     cv2.line(tracking_img, uv, (us[i+1], vs[i+1]), (255, 150, 0), 2)
    
    # tracking_img_msg = ros_numpy.msgify(Image, tracking_img, 'rgb8')
    # tracking_img_pub.publish(tracking_img_msg)

    # publish mask
    mask_img_msg = ros_numpy.msgify(Image, mask, 'rgb8')
    mask_img_pub.publish(mask_img_msg)


if __name__=='__main__':
    rospy.init_node('test', anonymous=True)

    rgb_sub = message_filters.Subscriber('/camera/color/image_raw', Image)
    depth_sub = message_filters.Subscriber('/camera/aligned_depth_to_color/image_raw', Image)
    pc_sub = message_filters.Subscriber('/camera/depth/color/points', PointCloud2)

    # header
    header = std_msgs.msg.Header()
    header.stamp = rospy.Time.now()
    header.frame_id = 'camera_color_optical_frame'
    fields = [PointField('x', 0, PointField.FLOAT32, 1),
                PointField('y', 4, PointField.FLOAT32, 1),
                PointField('z', 8, PointField.FLOAT32, 1),
                PointField('rgba', 12, PointField.UINT32, 1)]
    pc_pub = rospy.Publisher ('/pts', PointCloud2, queue_size=10)
    init_nodes_pub = rospy.Publisher ('/init_nodes', PointCloud2, queue_size=10)
    nodes_pub = rospy.Publisher ('/nodes', PointCloud2, queue_size=10)
    guide_nodes_pub = rospy.Publisher ('/guide_nodes', PointCloud2, queue_size=10)
    tracking_img_pub = rospy.Publisher ('/tracking_img', Image, queue_size=10)
    mask_img_pub = rospy.Publisher('/mask', Image, queue_size=10)

    ts = message_filters.TimeSynchronizer([rgb_sub, depth_sub, pc_sub], 10)
    ts.registerCallback(callback)

    rospy.spin()

File: src/cpp/mask.cpp
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <opencv2/highgui/highgui.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/features2d.hpp>

#include <message_filters/subscriber.h>
#include <message_filters/time_synchronizer.h>

#include <sensor_msgs/PointCloud2.h>
#include <pcl_conversions/pcl_conversions.h>
#include <pcl/point_types.h>
#include <pcl_ros/point_cloud.h>
#include <pcl/filters/conditional_removal.h>
#include <pcl/io/pcd_io.h>

using cv::Mat;

sensor_msgs::ImagePtr imageCallback(const sensor_msgs::ImageConstPtr& msg, const sensor_msgs::PointCloud2ConstPtr& _) {
    std::vector<int> lower_blue = {90, 60, 40};
    std::vector<int> upper_blue = {130, 255, 255};

    std::vector<int> lower_red_1 = {130, 60, 40};
    std::vector<int> upper_red_1 = {255, 255, 255};

    std::vector<int> lower_red_2 = {0, 60, 40};
    std::vector<int> upper_red_2 = {10, 255, 255};

    sensor_msgs::ImagePtr mask_msg = nullptr;

    try {
        Mat cur_image = cv_bridge::toCvShare(msg, "bgr8")->image;
        Mat cur_image_hsv;
        Mat mask_blue, mask_red_1, mask_red_2, mask_red, mask, mask_rgb;

        // convert color
        cv::cvtColor(cur_image, cur_image_hsv, cv::COLOR_BGR2HSV);

        // filter blue
        cv::inRange(cur_image_hsv, cv::Scalar(lower_blue[0], lower_blue[1], lower_blue[2]), cv::Scalar(upper_blue[0], upper_blue[1], upper_blue[2]), mask_blue);

        // filter red
        cv::inRange(cur_image_hsv, cv::Scalar(lower_red_1[0], lower_red_1[1], lower_red_1[2]), cv::Scalar(upper_red_1[0], upper_red_1[1], upper_red_1[2]), mask_red_1);
        cv::inRange(cur_image_hsv, cv::Scalar(lower_red_2[0], lower_red_2[1], lower_red_2[2]), cv::Scalar(upper_red_2[0], upper_red_2[1], upper_red_2[2]), mask_red_2);

        // combine red mask
        cv::bitwise_or(mask_red_1, mask_red_2, mask_red);
        // combine overall mask
        cv::bitwise_or(mask_red, mask_blue, mask);

        // simple blob detector
        cv::SimpleBlobDetector::Params blob_params;
        blob_params.filterByColor = false;
        blob_params.filterByArea = true;
        blob_params.filterByCircularity = false;
        blob_params.filterByInertia = true;
        blob_params.filterByConvexity = false;
        cv::Ptr<cv::SimpleBlobDetector> detector = cv::SimpleBlobDetector::create(blob_params);
        // detect
        std::vector<cv::KeyPoint> keypoints;
        detector->detect(mask_red, keypoints);

        cv::cvtColor(mask, mask_rgb, cv::COLOR_GRAY2BGR);

        // draw points
        for (cv::KeyPoint key_point : keypoints) {
            cv::circle(mask_rgb, key_point.pt, 5, cv::Scalar(0, 150, 255), -1);
        }

        // cv::imshow("view", mask_rgb);
        // cv::waitKey(30);

        // publish image
        mask_msg = cv_bridge::CvImage(std_msgs::Header(), "bgr8", mask_rgb).toImageMsg();
    }
    catch (cv_bridge::Exception& e) {
        ROS_ERROR("Could not convert from '%s' to 'bgr8'.", msg->encoding.c_str());
    }

    return mask_msg;
}

int main(int argc, char **argv) {
    ros::init(argc, argv, "image_listener");
    ros::NodeHandle nh;
    cv::namedWindow("view");

    image_transport::ImageTransport it(nh);
    image_transport::Publisher mask_pub = it.advertise("/mask", 1);

    // image_transport::Subscriber sub = it.subscribe("/camera/color/image_raw", 1, [&](const sensor_msgs::ImageConstPtr& msg){
    //     sensor_msgs::ImagePtr test_image = imageCallback(msg);
    //     mask_pub.publish(test_image);
    // });

    message_filters::Subscriber<sensor_msgs::Image> image_sub(nh, "/camera/color/image_raw", 1);
    message_filters::Subscriber<sensor_msgs::PointCloud2> pc_sub(nh, "/camera/depth/color/points", 1);
    message_filters::TimeSynchronizer<sensor_msgs::Image, sensor_msgs::PointCloud2> sync(image_sub, pc_sub, 1);

    sync.registerCallback<std::function<void(const sensor_msgs::ImageConstPtr&, 
                                             const sensor_msgs::PointCloud2ConstPtr&,
                                             const boost::shared_ptr<const message_filters::NullType>,
                                             const boost::shared_ptr<const message_filters::NullType>,
                                             const boost::shared_ptr<const message_filters::NullType>,
                                             const boost::shared_ptr<const message_filters::NullType>,
                                             const boost::shared_ptr<const message_filters::NullType>,
                                             const boost::shared_ptr<const message_filters::NullType>,
                                             const boost::shared_ptr<const message_filters::NullType>)>>
    (
        [&](const sensor_msgs::ImageConstPtr& msg, 
            const sensor_msgs::PointCloud2ConstPtr& _,
            const boost::shared_ptr<const message_filters::NullType> var1,
            const boost::shared_ptr<const message_filters::NullType> var2,
            const boost::shared_ptr<const message_filters::NullType> var3,
            const boost::shared_ptr<const message_filters::NullType> var4,
            const boost::shared_ptr<const message_filters::NullType> var5,
            const boost::shared_ptr<const message_filters::NullType> var6,
            const boost::shared_ptr<const message_filters::NullType> var7)
        {
            sensor_msgs::ImagePtr test_image = imageCallback(msg, _);
            mask_pub.publish(test_image);
        }
    );
    
    ros::spin();
    cv::destroyWindow("view");
}